codex/sales/salesagent.nim

```diff
@@ -21,6 +21,10 @@ type
     context*: SalesContext
     data*: SalesData
     subscribed: bool
+    onCleanUp*: OnCleanUp
+
+  OnCleanUp* = proc: Future[void] {.gcsafe, upraises: [].}
```

Comment:
Took me a while to remember, thank god for well-documented git comments 😄 

You're right, this should not be duplicated, which has been updated (by removing `OnCleanUp` from `SalesContext`.

The reason this was moved was because every time `processSlot` was called, the context's `onCleanUp` callback was overwritten, taking the `done` future in its closure with it. Effectively, `done` was only called for the *last* processed slot, and the other workers would remain blocked. This will likely also happen with `onFilled`, so I've moved that to `SalesAgent`. 

We have slot/agent-level callbacks, that are specific to the slot and the callback closure will be overwritten for each slot that is process. Then we have sales-level callbacks, in which their closure will not be overwritten for each slot processed. I've added som comments regarded this.

`OnSale` is a sales-level callback, which is triggered once a slot is filled. It is only used in the tests, which I don't particularly care for, but I don't think we have better options for testing that `filled` is called at this point.

Reply:
See https://github.com/codex-storage/nim-codex/pull/535/commits/df7ac216a01f3d0bae2effb4c1cc53b80c0403f2 for all the above changes.